refactor(provider): to make fields required and remove default values

emmenko opened the pull request:
commercetools now officially supports regions and cloud providers in the hostnames.

https://docs.commercetools.com/http-api.html#regions

~~Therefore, we should prefer to use `region` and `cloud_provider` fields in the TF provider and let the provider construct the correct hostname.~~ Also, there should not be any default values, as the user should explicitly provide them.

~~The fields `token_url` and `api_url` have been marked as `Removed` in the TF provider schema. However, the `CTP_API_URL` and `CTP_AUTH_URL` environment variables can still be used, for example for the test environment.~~

Update: this PR now changes the following:
- all the fields must be required
- there should not be any default values

emmenko commented on commercetools/provider.go:
```diff
@@ -17,52 +18,57 @@ func Provider() terraform.ResourceProvider {
 			"client_id": {
-				Type:     schema.TypeString,
-				Optional: true,
-				DefaultFunc: schema.MultiEnvDefaultFunc([]string{
```
I couldn't find a reason for using `MultiEnvDefaultFunc`, so I opted to use `EnvDefaultFunc` (same for the other fields)

davidweterings replied:
Makes sense :+1:

emmenko commented on commercetools/provider.go:
```diff
@@ -17,52 +18,57 @@ func Provider() terraform.ResourceProvider {
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("CTP_CLIENT_ID", nil),
+				Description: "The OAuth Client ID for a commercetools platform project. https://docs.commercetools.com/http-api-authorization",
+				Sensitive:   true,
```
<img width="655" alt="image" src="https://user-images.githubusercontent.com/1110551/75192573-c314a180-5754-11ea-9c35-ed63a2ddf3d1.png">

emmenko commented on commercetools/provider.go:
```diff
@@ -17,52 +18,57 @@ func Provider() terraform.ResourceProvider {
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("CTP_API_URL", nil),
+				Description: "The API URL of the commercetools platform. https://docs.commercetools.com/http-api",
+				Removed:     "Use the region and cloud_provider fields, to let the provider construct the correct hostname.",
```
Here the field is marked as `Removed`

emmenko commented on commercetools/provider.go:
```diff
@@ -17,52 +18,57 @@ func Provider() terraform.ResourceProvider {
+			},
+			"region": {
+				Type:        schema.TypeString,
+				Required:    true,
```
The new fields `region` and `cloud_provider` should be required.

davidweterings replied:
Nice! I'm wondering if we should validate the actual values, there seem to be a limited set of options right now: https://docs.commercetools.com/http-api#hosts

emmenko replied:
The `region` and `cloud_provider` have been removed again. See https://github.com/labd/terraform-provider-commercetools/pull/107#issuecomment-590607129

emmenko replied:
We could look into those fields again as a follow up maybe.

emmenko commented on commercetools/provider.go:
```diff
@@ -86,27 +92,26 @@ func Provider() terraform.ResourceProvider {
 	scopesRaw := d.Get("scopes").(string)
-	var scopes []string
-	if scopesRaw == "" {
-		scopes = []string{fmt.Sprintf("manage_project:%s", projectKey)}
```
The `manage_project` scope is [not recommended to be used](https://docs.commercetools.com/http-api-scopes#manage_projectprojectkey)

<img width="809" alt="image" src="https://user-images.githubusercontent.com/1110551/75192822-52ba5000-5755-11ea-8a06-66848cff972c.png">

Therefore, we should not provide it as a default value. The user must specify the scopes explicitly.

cneijenhuis commented:
Sorry, I may simply be unaware of an upcoming change so excuse the question:

CTP_API_URL and CTP_AUTH_URL are the current standard when creating an API client in the MC (both copy-able  and exportable in e.g. the environment variables template) and also currently used by the SDKs.

Is there a plan to change that? Otherwise this proposal does make it a bit more difficult to setup the config.

emmenko commented:
Thanks for the feedback. I just thought it would be easier to use the region/cloud_provider as the new hostnames follow a predictable format and thus allowing to potentially build any platform URL without having the user to add them manually. Other cloud providers also have similar fields.

However, I can see that those env variables are used by SDKs etc and it would be a bit inconvenient to change that. I'll remove the fields again.

emmenko commented:
Btw, to clarify some of the reasoning for having fields like `region` and `cloud_provider`, is to help avoiding possible mistakes. For example, I can potentially set `CTP_AUTH_URL` to one domain and `CTP_API_URL` to another domain. Other resources might also need other APIs, for example for managing Custom Applications we need to specify the MC API URL as well.

Having to manually define those values can potentially lead to using URLs from different regions. Of course it would be silly to do so but still it would be possible.

On the other hand, having to specify only `region` and `cloud_provider`, the URLs would be constructed by the provider and its resources, without any chance of mistakenly put the wrong values.

Just something to keep in mind for future improvements maybe.

emmenko commented on .azure-pipelines/jobs/test-ubuntu.yml:
```diff
@@ -24,11 +24,12 @@ jobs:
         env:
           TF_ACC: 1
-          CTP_AUTH_URL: http://localhost:8989
-          CTP_PROJECT_KEY: unittest
```
Are the values `unittest` and `x` just random values or do they need to be like this?

davidweterings replied:
They need to be like this, it's running a docker container with this server: https://github.com/labd/commercetools-python-sdk/blob/master/src/commercetools/testing/server.py 

This way we can test without talking to an actual Commercetools instance.

emmenko replied:
Ok, I changed the test values back to the original form.

emmenko commented on .azure-pipelines/jobs/test-ubuntu.yml:
```diff
@@ -24,11 +24,12 @@ jobs:
+          CTP_CLIENT_ID: clientid
+          CTP_CLIENT_SECRET: clientsecret
+          CTP_PROJECT_KEY: projectkey
+          CTP_SCOPES: view_project:projectkey
```
I still think that the scopes must be provided, there should not be a default to `manage_project`. Can someone confirm that this is ok to change?

davidweterings replied:
yes, if it's not recommended for production we should indeed not offer it default like this.

emmenko commented on commercetools/provider.go:
```diff
@@ -17,52 +17,43 @@ func Provider() terraform.ResourceProvider {
-				}, nil),
-				Description: "CommercesTools Client ID",
+				Type:        schema.TypeString,
+				Required:    true,
```
All the values in the provider should be required as far as I know.

emmenko commented:
Alright, I changed a bit the scope of the PR. Please let me know if the intended changes are ok or if there are potential issues with that.

Also, I can't tell why the tests are failing. Is it because I changed a bit the values for the environment variables?
